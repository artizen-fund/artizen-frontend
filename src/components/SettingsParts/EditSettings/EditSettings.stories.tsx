import EditSettings from './'

export default {
  title: 'components/EditSettings',
  component: EditSettings,
  argTypes: {},
}

export const EditSettingsComponent = (props: any) => <EditSettings {...props} />
