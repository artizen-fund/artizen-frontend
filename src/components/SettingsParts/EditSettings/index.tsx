import { useState, useEffect } from 'react'
import styled from 'styled-components'
import { useMutation, useReactiveVar } from '@apollo/client'
import { UPDATE_USER } from '@gql'
import { Form, Button, SettingsFormHeader, Spinner } from '@components'
import { loggedInUserVar } from '@lib'
import { breakpoint, typography } from '@theme'
import { schema, uischema, initialState, FormState } from '@forms/editSettings'

const EditSettings = () => {
  const [updateUser] = useMutation(UPDATE_USER)
  const loggedInUser = useReactiveVar(loggedInUserVar)

  const [data, setData] = useState<FormState>()
  useEffect(() => {
    // set initialState
    if (!!data || !loggedInUser) return
    setData({
      ...initialState,
      artizenHandle: loggedInUser?.artizenHandle || initialState.artizenHandle,
      email: loggedInUser?.email || initialState.email,
    })
  }, [loggedInUser])

  const [processing, setProcessing] = useState(false)
  // todo: replace processing with [loading] from useMutation

  const saveChanges = async () => {
    if (!loggedInUser) return
    setProcessing(true)
    await updateUser({ variables: { ...loggedInUser, ...data } })
    setProcessing(false)
  }

  return !loggedInUser || !data ? (
    <Spinner minHeight="85vh" />
  ) : (
    <Wrapper>
      <SettingsFormHeader
        imgPath="/assets/illustrations/settings/profile.png"
        darkImgPath="/assets/illustrations/settings/profile-dark.png"
        title="Account Settings"
        subtitle="Check your settings below."
      />
      <FormWrapper>
        <Form {...{ schema, uischema, initialState, data, setData }} readonly={processing}>
          <StyledButton disabled={processing} onClick={() => saveChanges()} stretch level={0}>
            Save Changes
          </StyledButton>
        </Form>
      </FormWrapper>
    </Wrapper>
  )
}

const Wrapper = styled.div``

const FormWrapper = styled.div`
  display: grid;
  gap: 10px;
  grid-template-areas:
    'artizenHandle'
    'phoneNumber'
    'saveChanges';

  @media only screen and (min-width: ${breakpoint.desktop}px) {
    gap: 16px;
  }

  .vertical-layout,
  .vertical-layout-item {
    display: contents;
  }

  *[id='#/properties/artizenHandle'] {
    grid-area: artizenHandle;
  }

  *[id='#/properties/email'] {
    grid-area: email;
  }

  *[id='#/properties/phoneNumber'] {
    grid-area: phoneNumber;
  }
`

const StyledButton = styled(props => <Button {...props} />)`
  grid-area: saveChanges;
  margin-top: 20px;
`

export default EditSettings
