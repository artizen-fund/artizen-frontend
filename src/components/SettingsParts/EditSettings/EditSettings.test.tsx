import renderer from 'react-test-renderer'
import EditSettings from './'

describe('EditSettings', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<EditSettings />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
