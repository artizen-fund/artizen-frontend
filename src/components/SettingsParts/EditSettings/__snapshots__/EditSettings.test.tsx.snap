// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditSettings renders correctly 1`] = `
<div
  className="sc-4d070d3e-0"
>
  <div
    className="vertical-layout"
    hidden={false}
  >
    <div
      className="vertical-layout-item"
    >
      <div
        className="sc-2f9d5146-0 imlAYL"
        id="#/properties/someString"
      >
        <div
          className="sc-dccedff6-0 jjRMSF"
        >
          <input
            defaultValue="Some default value"
            minLength={2}
            onBlur={[Function]}
            onChange={[Function]}
            placeholder=" "
            required={false}
            type="text"
          />
          <label
            className="sc-ffdb2e6c-0 cqczZc"
          >
            Tell us a thing.
          </label>
        </div>
        <div
          className="sc-93ba83c0-0 BZJQU"
        />
      </div>
    </div>
    <div
      className="vertical-layout-item"
    >
      <div
        className="sc-2f9d5146-0 eQAzBx"
        id="#/properties/someEmail"
      >
        <div
          className="sc-dccedff6-0 jjRMSF"
        >
          <input
            onBlur={[Function]}
            onChange={[Function]}
            placeholder="your@email.com"
            required={true}
            type="text"
          />
          <label
            className="sc-ffdb2e6c-0 cqczZc"
          >
            Email Address
             *
          </label>
        </div>
        <div
          className="sc-93ba83c0-0 BZJQU"
        />
      </div>
    </div>
    <div
      className="vertical-layout-item"
    >
      <div
        className="sc-2f9d5146-0 cKQxBV"
        id="#/properties/someNumber"
      >
        <div
          className="sc-dccedff6-0 jjRMSF"
        >
          <input
            className="noData"
            onBlur={[Function]}
            onChange={[Function]}
            placeholder=" "
            required={false}
            step="any"
            type="number"
          />
          <label
            className="sc-ffdb2e6c-0 cqczZc"
          >
            What is your lucky number?
          </label>
        </div>
        <div
          className="sc-93ba83c0-0 BZJQU"
        />
      </div>
    </div>
    <div
      className="vertical-layout-item"
    >
      <label
        className="sc-dd35d48e-4 hfvlkX"
        id="#/properties/someBoolean"
      >
        <div
          className="sc-dd35d48e-0 bkkyVw"
        >
          <input
            checked={true}
            className="sc-dd35d48e-1 clTMzK"
            disabled={false}
            onChange={[Function]}
            required={true}
            type="checkbox"
          />
          <span
            className="sc-dd35d48e-2 Legqi"
          />
        </div>
        <span
          className="sc-dd35d48e-3 kiyWfS"
        >
          Do you like to click checkmarks?
        </span>
      </label>
    </div>
  </div>
  <button
    className="sc-2a588df4-2 itVOtl undefined "
    onClick={[Function]}
  >
    <span>
      Submit
    </span>
  </button>
</div>
`;
